### ISSUE TITLE
M71 AudioDeviceModuleImpl says current platform is not supported (UWP)

### ISSUE THREAD

**Issue by Holo-Krzysztof** [webrtc-uwp/webrtc-uwp-sdk#104]
We're currently working on a D3D11 frame source for video. Our work is based on the `unityplugin` target because it's easy to integrate into all sorts of applications; however, when calling `CreatePeerConnection` the app crashes because webrtc calls abort inside the adm creation code.
Here's the log up to the point where it crashes:
```
(audio_processing_impl.cc:423): Capture analyzer activated: 0
Capture post processor activated: 0
Render pre processor activated: 0
(webrtcvoiceengine.cc:204): WebRtcVoiceEngine::WebRtcVoiceEngine
(webrtcvideoengine.cc:482): WebRtcVideoEngine::WebRtcVideoEngine()
(webrtcvoiceengine.cc:227): WebRtcVoiceEngine::Init
(webrtcvoiceengine.cc:234): Supported send codecs in order of preference:
(webrtcvoiceengine.cc:237): opus/48000/2 { minptime=10 useinbandfec=1 } (111)
(webrtcvoiceengine.cc:237): ISAC/16000/1 (103)
(webrtcvoiceengine.cc:237): ISAC/32000/1 (104)
(webrtcvoiceengine.cc:237): G722/8000/1 (9)
(webrtcvoiceengine.cc:237): ILBC/8000/1 (102)
(webrtcvoiceengine.cc:237): PCMU/8000/1 (0)
(webrtcvoiceengine.cc:237): PCMA/8000/1 (8)
(webrtcvoiceengine.cc:237): CN/32000/1 (106)
(webrtcvoiceengine.cc:237): CN/16000/1 (105)
(webrtcvoiceengine.cc:237): CN/8000/1 (13)
(webrtcvoiceengine.cc:237): telephone-event/48000/1 (110)
(webrtcvoiceengine.cc:237): telephone-event/32000/1 (112)
(webrtcvoiceengine.cc:237): telephone-event/16000/1 (113)
(webrtcvoiceengine.cc:237): telephone-event/8000/1 (126)
(webrtcvoiceengine.cc:240): Supported recv codecs in order of preference:
(webrtcvoiceengine.cc:243): opus/48000/2 { minptime=10 useinbandfec=1 } (111)
(webrtcvoiceengine.cc:243): ISAC/16000/1 (103)
(webrtcvoiceengine.cc:243): ISAC/32000/1 (104)
(webrtcvoiceengine.cc:243): G722/8000/1 (9)
(webrtcvoiceengine.cc:243): ILBC/8000/1 (102)
(webrtcvoiceengine.cc:243): PCMU/8000/1 (0)
(webrtcvoiceengine.cc:243): PCMA/8000/1 (8)
(webrtcvoiceengine.cc:243): CN/32000/1 (106)
(webrtcvoiceengine.cc:243): CN/16000/1 (105)
(webrtcvoiceengine.cc:243): CN/8000/1 (13)
(webrtcvoiceengine.cc:243): telephone-event/48000/1 (110)
(webrtcvoiceengine.cc:243): telephone-event/32000/1 (112)
(webrtcvoiceengine.cc:243): telephone-event/16000/1 (113)
(webrtcvoiceengine.cc:243): telephone-event/8000/1 (126)
(audio_device_impl.cc:73): Create
(audio_device_impl.cc:80): CreateForTest
(audio_device_buffer.cc:64): AudioDeviceBuffer::ctor
(audio_device_impl.cc:123): AudioDeviceModuleImpl
(audio_device_impl.cc:127): CheckPlatform
(audio_device_impl.cc:147): current platform is not supported => this module will self destruct!
(audio_device_impl.cc:318): ~AudioDeviceModuleImpl
(audio_device_buffer.cc:76): AudioDeviceBuffer::~dtor
```

I've skimmed the `audio_device_impl.cc` file and UWP support seems to have been removed (I remember this working in M66, I believe). What's the reason behind this decision and what's the plan for the future?

Also, for now we only need video to test. I've seen many overloads for `CreatePeerConnection`, but do you know if there is one that allows us to disable audio only leaving the rest as it is?

**Comment by Holo-Krzysztof** [comment]
Update: managed to get past the crash by constructing a `webrtc::FakeAudioDeviceModule` and passing that to `CreatePeerConnectionFactory`.

Having audio some time down the road would still be nice, though. :)

**Comment by jamescadd** [comment]
How recent is your clone?  There were some audio PRs yesterday though I'm not sure if it's completely working in releases/m71
https://github.com/webrtc-uwp/webrtc-uwp-sdk/pull/103

**Comment by ArmsOfSorrow** [comment]
Yeah, I saw that but that's in the sdk/wrapper layer which we don't use (unless it's easy to pick just the adm implementation without including everything else).
When trying the unityplugin dll it crashes at runtime because there is no adm implementation for UWP hooked up as default (like WASAPI was in M66, but it seems to have been deleted since).

**Comment by Holo-Krzysztof** [comment]
Yeah, I saw that but that's in the sdk/wrapper layer which we don't use (unless it's easy to pick just the adm implementation without including everything else).
When trying the unityplugin dll it crashes at runtime because there is no adm implementation for UWP hooked up as default (like WASAPI was in M66, but it seems to have been deleted since).

**Comment by morosev** [comment]
@Holo-Krzysztof Yes, audio device has been moved to external implementation in M71 and it is part of the wrapper code. Also, it depends on some wrapper components and it cannot be used out of wrapper project. On the other hand, only minor parts of the audio device in M71 are related to the wrapper dependencies. The audio module could be easily tailored to behave as a stand alone component. As an independent module it could be attached to the `CreatePeerConnectionFactory` without wrapper code.

**Comment by Holo-Krzysztof** [comment]
It'd be awesome if it worked as an independent ADM with `CreatePeerConnectionFactory`.
Are you guys planning to enable this? We're currently focused on getting video working; actually we got our first image rendered on a client made from d3d textures today, which is pretty nice. For now it's still very basic, i.e. we copy the existing frame to a staging texture and convert/encode on the CPU, but in the future we want to keep the texture on the GPU as long as possible in addition to using MF/D3D for encoding.

I'd still have to talk to the higher-ups about this at some point, but assuming they're okay with open-sourcing, would you be interested in merging our D3D Frame Source into your repository?

**Comment by jamescadd** [comment]
I wanted to better understand if you need unmanaged access to the ADM, and if so why.  Would it work to be able to use the current wrapper class if the ADM could be instantiated independently of other external modules?

**Comment by Holo-Krzysztof** [comment]
Sorry, I didn't check my mails on time and missed the invite to your call. :/

Generally, we'd like to share as much code as possible across platforms with the major piece being
a DLL exposing a C API for creating `PeerConnection`/`DataChannel` that uses libwebrtc C++ API internally (basically the `examples/unityplugin` target evolved into something production-ready and not Unity-dependent is our goal). 

As @morosev pointed out, we're using `CreatePeerConnectionFactory` for initialization, so using the existing ADM would work if it were independent (assuming it behaves well without the rest of the wrapper code being around). Note that I haven't read very much of the wrapper code, so it's just speculation/wishful thinking on my part.

About our audio needs, we'll need the ability to capture audio played by an application instead of the microphone at some point. Does the current ADM support that scenario? If not, would it be hard to integrate?

**Comment by jamescadd** [comment]
Np, it was a last minute call :)  Our current thinking is that the ADM wrapper code should work independently though it's based on a brief review.  I would recommend starting with this approach and reaching out with any issues.  @morosev may have some guidance on using audio from the application, I believe the ADM has a native plugin interface which we have considered exposing to managed code that would allow this.  Not sure how much that has changed with the new ADM in M72 though (not in our codebase yet).

**Comment by ispysoftware** [comment]
We desperately need a way to push raw audio byte arrays into the wrapper to send out to clients as part of the audio stream. We're taking audio from ip cameras and network sources and need a way to send that out.
